### PR TITLE
1.x: Merging an observable of singles.

### PR DESCRIPTION
**Pull request by abersnaze**
The addition of a `Observable<Single<T>> -> Observable<T>` to round out the basic API of `rx.Single`. I need this for doing a flat scan of sorts.

```
Single.merge(events.scan(Single.just(seed), (stateSingle, event) -> {
    return stateSingle.flatMap((state) -> {
        return state.write(event);
    }).cache();
}));
```

### CODE REVIEW

**Comment by davidmoten** [comment]
Looks like a big copy and paste from an existing merge operator. Can you summarize the differences from the original tried and true operator to help review? I also wonder if we can share a decent chunk of code between operators.

**Comment by abersnaze** [comment]
It was a copy for the most part but for changing the class names. It could have be done by mapping the `Observable<Single<T>>` to an `Observable<Observable<T>>`.

File: src/main/java/rx/Single.java [hunk 1]
```diff
@@ -931,6 +931,110 @@ public void onError(Throwable error) {
...
+     * </dl>
+     *
+     * @param <T> the common value type
+     * @param source
+     *            an Observable of Singles to be merged
+     * @return an Observable that emits all of the items emitted by the source Singles
+     * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
+     */
+    public static <T> Observable<T> merge(Observable<? extends Single<T>> source) {
+        return source.lift(OperatorMergeSingle.<T> instance(false));
```
Comment (akarnokd) [inline src/main/java/rx/Single.java:952]: This could be

```java
return source.map(new Func1<Single<T>, Observable<T>() {
    @Override public Observable<T> call(Single<T> t) {
        return t.toObservable();
    }
}).lift(OperatorMerge.<T>instance(false));
```

File: src/main/java/rx/Single.java [hunk 2]
```diff
@@ -931,6 +931,110 @@ public void onError(Throwable error) {
...
+     * <dd>{@code merge} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     *
+     * @param <T> the common value type
+     * @param source
+     *            an Observable of Singles to be merged
+     * @return an Observable that emits all of the items emitted by the source Singles
+     * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
+     */
+    public static <T> Observable<T> merge(Observable<? extends Single<T>> source) {
```
Comment (akarnokd) [inline src/main/java/rx/Single.java:951]: Experimental

File: src/main/java/rx/Single.java [hunk 3]
```diff
@@ -931,6 +931,110 @@ public void onError(Throwable error) {
...
+     *
+     * @param <T> the common value type
+     * @param source
+     *            an Observable of Singles to be merged
+     * @param maxConcurrent
+     *            the maximum number of Singles that may be subscribed to concurrently
+     * @return an Observable that emits all of the items emitted by the source Singles
+     * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
+     */
+    public static <T> Observable<T> merge(Observable<? extends Single<T>> source, int maxConcurrent) {
```
Comment (akarnokd) [inline src/main/java/rx/Single.java:975]: Experimental

File: src/main/java/rx/Single.java [hunk 4]
```diff
@@ -931,6 +931,110 @@ public void onError(Throwable error) {
...
+     * <dd>{@code merge} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     *
+     * @param <T> the common value type
+     * @param source
+     *            an Observable of Singles to be merged
+     * @return an Observable that emits all of the items emitted by the source Singles
+     * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
+     */
+    public static <T> Observable<T> mergeDelayError(Observable<? extends Single<T>> source) {
```
Comment (akarnokd) [inline src/main/java/rx/Single.java:1003]: Experimental

File: src/main/java/rx/Single.java [hunk 5]
```diff
@@ -931,6 +931,110 @@ public void onError(Throwable error) {
...
+     *
+     * @param <T> the common value type
+     * @param source
+     *            an Observable of Singles to be merged
+     * @param maxConcurrent
+     *            the maximum number of Singles that may be subscribed to concurrently
+     * @return an Observable that emits all of the items emitted by the source Singles
+     * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
+     */
+    public static <T> Observable<T> mergeDelayError(Observable<? extends Single<T>> source, int maxConcurrent) {
```
Comment (akarnokd) [inline src/main/java/rx/Single.java:1033]: Experimental

**Review by akarnokd — CHANGES_REQUESTED** [review]
Please reuse the existing merge operator.

**Comment by abersnaze** [comment]
@akarnokd actually do you think that the back pressure could be streamlined because of the guarantee that they'll be one onNext for each subscription? Does it even need a custom Producer if we got rid of the maxConcurrent overloads?

**Comment by akarnokd** [comment]
The [v2 version](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/flowable/FlowableFlatMapSingle.java) is optimized.

**Comment by akarnokd** [comment]
> if we got rid of the maxConcurrent overloads

You may not want to have a million outstanding `Single`-based network request. Limiting flatMap's concurrency is expected here where backpressure helps naturally.

**Comment by akarnokd** [comment]
Any progress on this?

**Comment by akarnokd** [comment]
Closing via #5092.
